toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -411,8 +421,66 @@ auto BuiltinFunctionKind::IsValidType(const File& sem_ir,
   return ValidateFns[AsInt()](sem_ir, arg_types, return_type);
 }
 
-auto BuiltinFunctionKind::IsCompTimeOnly() const -> bool {
-  return *this == BuiltinFunctionKind::IntConvertChecked;
+auto BuiltinFunctionKind::IsCompTimeOnly(const File& sem_ir,
+                                         llvm::ArrayRef<InstId> arg_ids,
+                                         TypeId return_type_id) const -> bool {
+  switch (*this) {
+    case IntConvertChecked:
+      // Checked integer conversions are compile-time only.
+      return true;
+
+    case IntSNegate:
+    case IntComplement:
+    case IntSAdd:
+    case IntSSub:
+    case IntSMul:
+    case IntSDiv:
+    case IntSMod:
+    case IntAnd:
+    case IntOr:
+    case IntXor:
+      // Integer builtins producing an IntLiteral are compile-time only.
+      // TODO: We could allow these at runtime and just produce an empty struct
+      // result. Should we?
+      return sem_ir.types().Is<SemIR::IntLiteralType>(return_type_id);
+
+    case IntLeftShift:
+    case IntRightShift:
+      // Shifts by an integer literal amount are compile-time only. We don't
+      // have a value for the shift amount at runtime in general.
+      // TODO: We could allow these in the case where the shift amount has a
+      // compile-time value.
```

Comment:
Given that the shift amount has to fit into the shifted operand's type (and LLVM even makes us use that!), why not support there?

Or maybe this is what you mean by "compile-time value" -- but don't integer literals *always* have a compile time value?

Sorry if the comment is just confusing me (see below as well).

Reply:
Values of integer literal type don't always have a compile-time value. Eg, given:
```
var n: Core.IntLiteral() = 123;
```
... we can't lower `(1 as i32) << n` because the runtime representation of `n` doesn't track its value at all. (We also can't convert `n` to another type; it ends up being pretty useless, because it only has compile-time operations, much like a runtime value of type `type`.)

Reply:
This comment actually helped me understand what all of these TODOs were really about more than the comments in the source code, so I think there is a remaining improvement available in the comments here.

Maybe a single comment above this switch that clarifies that we need to restrict operations on integer literals to be compile time only because at runtime we no longer have the value of the literal, and just know that it was a literal? You could even give the example that helps surface this. Then the TODOs here would I think make more sense, in particular "where the shift amount has a compile-time value" would connect back to that for me.

Reply:
Added a comment before the switch to try to explain this a bit better.